lib/connection.js

```diff
@@ -905,45 +905,62 @@ function _setClient(conn, client, options, dbName) {
  * Closes the connection
  *
  * @param {Boolean} [force] optional
+ * @param {Boolean} [closeAndClean] optional
  * @param {Function} [callback] optional
  * @return {Promise}
  * @api public
  */
 
-Connection.prototype.close = function(force, callback) {
+Connection.prototype.close = function(force, closeAndClean, callback) {
   if (typeof force === 'function') {
     callback = force;
     force = false;
   }
 
+  if (typeof closeAndClean === 'function') {
+    callback = closeAndClean;
+    closeAndClean = false;
+  }
+
   if (force != null && typeof force === 'object') {
     this.$wasForceClosed = !!force.force;
   } else {
     this.$wasForceClosed = !!force;
   }
 
+  if (closeAndClean !== null) {
+    closeAndClean = !!closeAndClean
+  }
+
   return promiseOrCallback(callback, cb => {
-    this._close(force, cb);
+    this._close(force, closeAndClean, cb);
   });
 };
 
 /**
  * Handles closing the connection
  *
  * @param {Boolean} force
+ * @param {Boolean} closeAndClean
  * @param {Function} callback
  * @api private
  */
-Connection.prototype._close = function(force, callback) {
+Connection.prototype._close = function(force, closeAndClean, callback) {
   const _this = this;
   const closeCalled = this._closeCalled;
   this._closeCalled = true;
   if (this.client != null) {
     this.client._closeCalled = true;
   }
 
+  let conn = this;
   switch (this.readyState) {
     case STATES.disconnected:
+      if (closeAndClean && this.base.connections.indexOf(conn) != -1) {
+        console.log("BASE CONNECTIONS COUNT BEFORE 1---------> ", this.base.connections.length)
```

Comment:
Please remove the experimental logs from the PR.

Reply:
@AbdelrahmanHafez Yes, I am testing out these changes, will remove once testing is done, before I mark this ready to merge.